agent/dns.go

```diff
@@ -1812,7 +1819,11 @@ func (d *DNSServer) handleRecurse(resp dns.ResponseWriter, req *dns.Msg) {
 	var r *dns.Msg
 	var rtt time.Duration
 	var err error
-	for _, recursor := range cfg.Recursors {
+	var start = int(cfg.Recursors.RoundRobinIndex)
+	if cfg.RecursorRotate {
+		cfg.Recursors.RoundRobinIndex += 1
+	}
+	for _, recursor := range append(cfg.Recursors.Addrs[start:], cfg.Recursors.Addrs[:start]...) {
```

Comment:
I think there may be a problems here. `cfg.Recursors.RoundRobinIndex` does not appear to be reset to 0 when it reaches the end of the list. I think this will panic after a few requests.

I believe `cfg.Recursors.RoundRobinIndex += 1` is not safe for concurrent use either, so this could return incorrect results when there are multiple handlers running at the same time in different goroutines.

Instead of round robin, would random selection be ok?  Something like this would remove the need to track the `Index`, and I think should also be safe for concurrent use because we only read concurrently, and the recursors list does not change.

```go
for _, idx := range rand.Perm(len(cfg.Recursors)) {
    recursor := cfg.Recursors[idx]
}
````

Reply:
@dnephin Doh! I forgot to modulo.

I am sure this is not safe for concurrent use either, but I believe something like this could work as well? What do you think.

`cfg.Recursors.RoundRobinIndex = (atomic.AddInt32(cfg.Recursors.RoundRobinIndex, 1)%len(cfg.Recursors))`

Either way, I like the suggestion because it does a good job of distributing the load. It adds complexity to increment the index and store it as an extra variable. 

Would something like this be good/safe golang style?
`for idx, random_idx := range rand.Perm(len(cfg.Recursors)) {
    recursor := cfg.Recursors[idx]
   if cfg. RecursorRotate {
      recursor = cfg.Recursors[random_idx]
  }
}`

Reply:
@dnephin any thoughts on the changes?